evalml/automl/automl_search.py

```diff
@@ -627,6 +636,27 @@ def __init__(
             custom_hyperparameters=custom_hyperparameters,
         )
 
+    def _catch_warnings(self, warning_list):
+        if len(warning_list) == len(self.allowed_pipelines) and len(warning_list) > 0:
```

Comment:
@bchen1116 I think we can simplify this implementation a bit if we raise our own custom warning that subclasses `UserWarning`. 

```python
class ParameterNotUsedWarning(UserWarning):

    def __init__(self, components):
        self.components = components

        msg = f"Parameters for components {components} will not be used to instantiate the pipeline since they don't appear in the pipeline"
        super().__init__(msg)
```

ComponentGraph now becomes:

```python
        diff = param_set.difference(set(self.component_instances.keys()))
        if len(diff):
            warnings.warn(ParameterNotUsedWarning(diff))
```

_catch_warnings now becomes:

```python
    def _catch_warnings(self, warning_list):
        if len(warning_list) == len(self.allowed_pipelines) and len(warning_list) > 0:
            # we find the value(s) that we must throw the warning for
            final_message = set([])
            for idx, msg in enumerate(warning_list):
                if idx == 0:
                    final_message = final_message.union(msg.message.components)
                else:
                    final_message.intersection(msg.message.components)
            warnings.warn(ParameterNotUsedWarning(list(final_message)))
```

I think this is cleaner because it defines an api for accessing the components we want to warn about rather than messing with the string. I worry that'll be harder to maintain if we ever decide to change the message. I think the implementation is easier to understand as well. We can also explicitly filter for our custom warning in `AutoMLSearch` which seems cleaner too.

Something like: 
```python
            with warnings.catch_warnings(record=True) as w:
                warnings.filterwarnings("always", category=ParameterNotUsedWarning)
```

Let me know what you think!

Reply:
This is a good idea! Yeah, I can change the impl here

Reply:
@freddyaboulton added the suggestion you raised. Thanks!

Reply:
Thanks @bchen1116 I will take a look